core/src/main/java/org/apache/calcite/rel/rules/FilterFlattenCorrelatedConditionRule.java

```diff
@@ -0,0 +1,134 @@
+ * condition in a simpler form that is more convenient for the decorrelation logic.
+ *
+ * Uncorrelated calls below a comparison operator are turned into input references by extracting the
+ * computation in a {@link org.apache.calcite.rel.core.Project} expression. An additional projection
+ * may be added on top of the new filter to retain expression equivalence.
+ *
+ * Sub-plan before
+ * <pre>
+ * LogicalProject($f0=[true])
+ *   LogicalFilter(condition=[=($cor0.DEPTNO, +($7, 30))])
+ *     LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ * </pre>
+ *
+ * Sub-plan after
+ * <pre>
+ * LogicalProject($f0=[true])
+ *   LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2],..., COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
+ *     LogicalFilter(condition=[=($cor0.DEPTNO, $9)])
+ *       LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2],..., SLACKER=[$8], $f9=[+($7, 30)])
+ *         LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ * </pre>
+ */
+public final class FilterFlattenCorrelatedConditionRule
+    extends RelRule<FilterFlattenCorrelatedConditionRule.Config> {
+
+  public FilterFlattenCorrelatedConditionRule(final Config config) {
+    super(config);
+  }
+
+  @Override public boolean matches(RelOptRuleCall call) {
+    Filter filter = call.rel(0);
+    return RexUtil.containsCorrelation(filter.getCondition());
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Filter filter = call.rel(0);
+    RelBuilder b = call.builder();
+    b.push(filter.getInput());
+    final int proj = b.fields().size();
+    List<RexNode> projOperands = new ArrayList<>();
+    RexNode newCondition = filter.getCondition().accept(new RexShuttle() {
+      @Override public RexNode visitCall(RexCall call) {
+        switch (call.getKind()) {
+        case EQUALS:
+        case NOT_EQUALS:
+        case GREATER_THAN:
+        case GREATER_THAN_OR_EQUAL:
+        case LESS_THAN:
+        case LESS_THAN_OR_EQUAL:
+          RexNode op0 = call.operands.get(0);
+          RexNode op1 = call.operands.get(1);
+          final int replaceIndex;
+          if (RexUtil.containsCorrelation(op1) && isUncorrelatedCall(op0)) {
+            replaceIndex = 0;
+          } else if (RexUtil.containsCorrelation(op0) && isUncorrelatedCall(op1)) {
+            replaceIndex = 1;
+          } else {
+            // Structure does not match, do not replace
+            replaceIndex = -1;
+          }
+          if (replaceIndex != -1) {
+            List<RexNode> copyOperands = new ArrayList<>(call.operands);
+            RexNode oldOp = call.operands.get(replaceIndex);
+            RexNode newOp = b.getRexBuilder()
+                .makeInputRef(oldOp.getType(), proj + projOperands.size());
+            projOperands.add(oldOp);
+            copyOperands.set(replaceIndex, newOp);
+            return call.clone(call.type, copyOperands);
+          }
+          return call;
+        case AND:
```

Comment:
I believe you also not want here.

Reply:
The rule is tightly connected with the current implementation in `RelDecorrelator#findCorrelationEquivalent`. If we cannot handle NOT there then we don't gain much by handling more kinds of expressions in this rule. I added some comments and extra javadoc in the code to better reflect the purpose.